File: packages/ai/src/utils/typebox-helpers.ts
import type { TSchema, TUnsafe } from "@sinclair/typebox";
import { Type } from "@sinclair/typebox";

/**
 * Creates a string enum schema compatible with Google's API and other providers
 * that don't support anyOf/const patterns.
 *
 * @example
 * const OperationSchema = StringEnum(["add", "subtract", "multiply", "divide"], {
 *   description: "The operation to perform"
 * });
 *
 * type Operation = Static<typeof OperationSchema>; // "add" | "subtract" | "multiply" | "divide"
 */
export function StringEnum<T extends readonly string[]>(
	values: T,
	options?: { description?: string; default?: T[number] },
): TUnsafe<T[number]> {
	return Type.Unsafe<T[number]>({
		type: "string",
		enum: values as any,
		...(options?.description && { description: options.description }),
		...(options?.default && { default: options.default }),
	});
}

/**
 * Validation error for tool schemas that are incompatible with Google's API.
 */
export interface SchemaValidationError {
	path: string;
	message: string;
}

/**
 * Validates a tool schema for Google API compatibility.
 *
 * Google's Generative AI API doesn't support certain JSON Schema features:
 * - `const` keyword (used by TypeBox for Type.Literal)
 * - `anyOf` with `const` values (used by TypeBox for Type.Union([Type.Literal(...)]))
 *
 * Use `StringEnum` instead of `Type.Union([Type.Literal(...)])` for string enums.
 *
 * @example
 * // In a test file:
 * import { validateToolSchema } from '@kennyfrc/mu-ai';
 * import { myTool } from './tools/my-tool.js';
 *
 * test('tool schema is Google-compatible', () => {
 *   const errors = validateToolSchema(myTool.parameters);
 *   expect(errors).toEqual([]);
 * });
 *
 * @returns Array of validation errors. Empty array if schema is valid.
 */
export function validateToolSchema(schema: TSchema, path = ""): SchemaValidationError[] {
	const errors: SchemaValidationError[] = [];

	if (typeof schema !== "object" || schema === null) {
		return errors;
	}

	const record = schema as Record<string, unknown>;

	// Check for 'const' keyword (Google doesn't support it)
	if ("const" in record) {
		errors.push({
			path: path || "(root)",
			message: `Found 'const' keyword which Google API doesn't support. Use StringEnum instead of Type.Literal.`,
		});
	}

	// Check for anyOf with const values (Type.Union([Type.Literal(...)]) pattern)
	if (Array.isArray(record.anyOf)) {
		const hasConst = record.anyOf.some((item) => typeof item === "object" && item !== null && "const" in item);
		if (hasConst) {
			errors.push({
				path: path || "(root)",
				message: `Found 'anyOf' with 'const' values which Google API doesn't support. Use StringEnum instead of Type.Union([Type.Literal(...)]).`,
			});
		}
		// Recurse into anyOf items
		for (let i = 0; i < record.anyOf.length; i++) {
			errors.push(...validateToolSchema(record.anyOf[i] as TSchema, `${path}.anyOf[${i}]`));
		}
	}

	// Recurse into properties
	if (typeof record.properties === "object" && record.properties !== null) {
		for (const [key, value] of Object.entries(record.properties)) {
			errors.push(...validateToolSchema(value as TSchema, `${path}.properties.${key}`));
		}
	}

	// Recurse into items (for arrays)
	if (typeof record.items === "object" && record.items !== null) {
		errors.push(...validateToolSchema(record.items as TSchema, `${path}.items`));
	}

	// Recurse into additionalProperties
	if (typeof record.additionalProperties === "object" && record.additionalProperties !== null) {
		errors.push(...validateToolSchema(record.additionalProperties as TSchema, `${path}.additionalProperties`));
	}

	return errors;
}

/**
 * Validates all tools in an array for Google API compatibility.
 *
 * @example
 * import { validateToolSchemas } from '@kennyfrc/mu-ai';
 * import { codingTools } from './tools/index.js';
 *
 * test('all tools are Google-compatible', () => {
 *   const errors = validateToolSchemas(codingTools);
 *   if (errors.length > 0) {
 *     console.error('Schema validation errors:', errors);
 *   }
 *   expect(errors).toEqual([]);
 * });
 *
 * @returns Array of validation errors with tool names. Empty array if all schemas are valid.
 */
export function validateToolSchemas(
	tools: Array<{ name: string; parameters: TSchema }>,
): Array<SchemaValidationError & { toolName: string }> {
	const allErrors: Array<SchemaValidationError & { toolName: string }> = [];

	for (const tool of tools) {
		const errors = validateToolSchema(tool.parameters);
		for (const error of errors) {
			allErrors.push({ ...error, toolName: tool.name });
		}
	}

	return allErrors;
}

File: python/src/pi_mono/ai/utils/typebox_helpers.py
"""TypeBox-like helpers for creating JSON schemas."""

from typing import Any, TypedDict


class StringEnumOptions(TypedDict, total=False):
    description: str
    default: str


def string_enum(
    values: list[str],
    options: StringEnumOptions | None = None,
) -> dict[str, Any]:
    """
    Creates a string enum schema compatible with Google's API and other providers
    that don't support anyOf/const patterns.

    Args:
        values: List of valid string values
        options: Optional dict with 'description' and/or 'default'

    Returns:
        JSON schema dict for a string enum

    Example:
        >>> OperationSchema = string_enum(
        ...     ["add", "subtract", "multiply", "divide"],
        ...     {"description": "The operation to perform"}
        ... )
    """
    schema: dict[str, Any] = {
        "type": "string",
        "enum": values,
    }
    if options:
        if options.get("description"):
            schema["description"] = options["description"]
        if options.get("default"):
            schema["default"] = options["default"]
    return schema


class NumberEnumOptions(TypedDict, total=False):
    description: str
    default: int | float


def number_enum(
    values: list[int | float],
    options: NumberEnumOptions | None = None,
) -> dict[str, Any]:
    """Creates a number enum schema."""
    schema: dict[str, Any] = {
        "type": "number",
        "enum": values,
    }
    if options:
        if options.get("description"):
            schema["description"] = options["description"]
        if options.get("default") is not None:
            schema["default"] = options["default"]
    return schema


def object_schema(
    properties: dict[str, dict[str, Any]],
    required: list[str] | None = None,
    additional_properties: bool | dict[str, Any] = False,
) -> dict[str, Any]:
    """Creates an object schema."""
    schema: dict[str, Any] = {
        "type": "object",
        "properties": properties,
    }
    if required:
        schema["required"] = required
    if additional_properties is not True:
        schema["additionalProperties"] = additional_properties
    return schema


def array_schema(
    items: dict[str, Any],
    min_items: int | None = None,
    max_items: int | None = None,
) -> dict[str, Any]:
    """Creates an array schema."""
    schema: dict[str, Any] = {
        "type": "array",
        "items": items,
    }
    if min_items is not None:
        schema["minItems"] = min_items
    if max_items is not None:
        schema["maxItems"] = max_items
    return schema
